File: back-product-mgmt/Infrastructure/Data/PostgresRepository.cs
using System.Text.Json;
using Core.Entities;
using Core.Interfaces;
using Core.Queries;
using Core.Results;
using Npgsql;

namespace Infrastructure.Data;

public class PostgresRepository : IProductRepository
{
    private readonly string _connectionString;

    private readonly JsonSerializerOptions _jsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };
    public PostgresRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task AddAsync(Product product)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        string productJson = JsonSerializer.Serialize(product, _jsonOptions);

        await using var command = new NpgsqlCommand(
            "INSERT INTO Products (Uuid, Data) VALUES (@uuid, @data::jsonb)", connection
        );
        command.Parameters.AddWithValue("uuid", product.Uuid);
        command.Parameters.AddWithValue("data", productJson);
        await command.ExecuteNonQueryAsync();
    }

    public async Task<bool> DeleteAsync(Guid uuid)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        await using var command = new NpgsqlCommand(
            "DELETE FROM Products WHERE Uuid = @uuid", connection
        );
        command.Parameters.AddWithValue("uuid", uuid);

        int deletedRows = await command.ExecuteNonQueryAsync();
        return deletedRows > 0;
    }

    public async Task<IEnumerable<Product>> GetAllAsync()
    {
        var products = new List<Product>();
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        await using var command = new NpgsqlCommand(
            "SELECT Data FROM Products", connection
        );
        await using var read = await command.ExecuteReaderAsync();
        while (await read.ReadAsync())
        {
            var product = JsonSerializer.Deserialize<Product>(read.GetString(0), _jsonOptions);
            if (product != null)
                products.Add(product);
        }
        return products;
    }

    public async Task<Product?> GetByIdAsync(Guid uuid)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        await using var command = new NpgsqlCommand(
            "SELECT Data FROM Products where Uuid = @uuid", connection
        );
        command.Parameters.AddWithValue("uuid", uuid);
        await using var read = await command.ExecuteReaderAsync();
        if (await read.ReadAsync())
        {
            var product = JsonSerializer.Deserialize<Product>(read.GetString(0), _jsonOptions);
            return product;
        }
        return null;
    }

    public async Task<Product?> UpdateAsync(Guid uuid, Product data)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        string productJson = JsonSerializer.Serialize(data, _jsonOptions);

        await using var command = new NpgsqlCommand(
            "UPDATE Products SET Data = @data::jsonb WHERE Uuid = @uuid", connection
        );
        command.Parameters.AddWithValue("uuid", uuid);
        command.Parameters.AddWithValue("data", productJson);
        int updatedRows = await command.ExecuteNonQueryAsync();
        if (updatedRows > 0)
        {
            return data;
        }
        return null;
    }

    public async Task<PagedResult<Product>> GetPagedAsync(ProductQuery query)
    {

        var products = new List<Product>();
        var sqlCount = "SELECT count(*) from Products WHERE 1=1";
        var filters = "";
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        await using var command = new NpgsqlCommand();

        //Name filters
        if (!string.IsNullOrEmpty(query.Name))
        {
            filters += " AND Data->>'Name' ILike @name";
            command.Parameters.AddWithValue("name", $"%{query.Name}%");
        }

        //ContainerId
        if (query.CategoryId.HasValue)
        {
            filters += " AND (Data->>'CategoryId')::int = @categoryId";
            command.Parameters.AddWithValue("categoryId", query.CategoryId);
        }

        //Get TotalCount
        command.CommandText = sqlCount + filters;
        command.Connection = connection;
        var totalCount = Convert.ToInt32(await command.ExecuteScalarAsync());

        if (totalCount == 0)
        {
            return new PagedResult<Product>(0, products);
        }

        //Sort
        var allowedSortColumns = new[] { "Name", "Price", "Quantity" };
        var sortColumn = allowedSortColumns.FirstOrDefault(c => c.Equals(query.OrderBy, StringComparison.OrdinalIgnoreCase)) ?? "Name";
        var direction = query.Descending ? "DESC" : "ASC";

        if (sortColumn.Equals("Price") || sortColumn.Equals("Quantity"))
        {
            filters += $" ORDER BY (Data->>'{sortColumn}')::numeric {direction}";
        }
        else
        {
            filters += $" ORDER BY Data->>'{sortColumn}' {direction}";
        }

        //Pagination
        filters += $" LIMIT @PageSize OFFSET @offset";
        command.Parameters.AddWithValue("PageSize", query.PageSize);
        command.Parameters.AddWithValue("offset", (query.PageNumber - 1) * query.PageSize);
        command.CommandText = $"SELECT Data FROM Products WHERE 1=1 {filters}";

        await using var read = await command.ExecuteReaderAsync();
        while (await read.ReadAsync())
        {
            var product = JsonSerializer.Deserialize<Product>(read.GetString(0), _jsonOptions);
            if (product != null)
                products.Add(product);
        }
        return new PagedResult<Product>(totalCount, products);
    }
}

File: back-order-mgmt/src/infrastructure/data/PostgresRepository.ts
import type { Pool } from "../../../node_modules/@types/pg/index.js";
import type { Order } from "../../core/Entities/Order.js";
import type{ IOrderRepository } from "../../core/Interfaces/IOrderRepository.js";
import type { OrderQuery } from "../../core/Queries/OrderQuery.js";
import type { PagedResult } from "../../core/Results/PagedResult.js";

export class PostgresRepository implements IOrderRepository {

    constructor (private readonly pool : Pool){};

    async getPagedAsync(query: OrderQuery): Promise<PagedResult<Order>> {
        try {
            const values: any[] = [];
            let filters = "";

            //filters
            if (query.customerId) {
                values.push(`%${query.customerId}%`);
                filters += ` AND Data->>'customerId' ILIKE $${values.length}`;
            }

            if (query.uuid) {
                values.push(`%${query.uuid}%`);
                filters += ` AND Uuid::text ILIKE $${values.length}`;
            }

            if (query.productIds && query.productIds.length > 0) {
                values.push(query.productIds);
                filters += ` AND EXISTS (
                    SELECT 1 FROM jsonb_array_elements(Data->'items') AS item 
                    WHERE item->>'productId' = ANY($${values.length})
                )`;
            }

            if (query.dateFrom) {
                values.push(query.dateFrom);
                filters += ` AND (Data->>'created_at')::timestamp >= $${values.length}::timestamp`;
            }

            if (query.dateTo) {
                values.push(query.dateTo);
                filters += ` AND (Data->>'created_at')::timestamp <= $${values.length}::timestamp`;
            }

            //totalCount
            const countSql = `SELECT count(*) FROM Orders WHERE 1=1 ${filters}`;
            const countResult = await this.pool.query(countSql, values);
            const totalCount = parseInt(countResult.rows[0].count, 10);

            if (totalCount === 0) {
                return { totalCount: 0, data: [] };
            }
            //Sort
            const allowedSortColumns = ["createdAt", "customerId"];
            const sortColumn = allowedSortColumns.includes(query.orderBy || "") ? query.orderBy : "createdAt";
            const direction = query.descending ? "DESC" : "ASC";

            let sortClause = "";
            if (sortColumn === "customerId") {
                sortClause = ` ORDER BY LOWER(Data->>'customerId') ${direction}`;
            } else {
                sortClause = ` ORDER BY (Data->>'created_at')::timestamp ${direction}`;
            }

            const dataValues = [...values];
            //offset
            dataValues.push(query.pageSize);
            const limitIndex = dataValues.length;

            const offset = (query.pageNumber - 1) * query.pageSize;
            dataValues.push(offset);
            const offsetIndex = dataValues.length;

            const paginationClause = ` LIMIT $${limitIndex} OFFSET $${offsetIndex}`;

            //execute
            const dataSql = `SELECT Data FROM Orders WHERE 1=1 ${filters} ${sortClause} ${paginationClause}`;
            const dataResult = await this.pool.query(dataSql, dataValues);

            const orders: Order[] = dataResult.rows.map(row => {
                const orderData = row.data;
                orderData.created_at = new Date(orderData.created_at);
                return orderData as Order;
            });

            return {
                totalCount,
                data: orders
            };

        } catch (error) {
            console.error("Error fetching orders by query from Postgres:", error);
            throw error;
        }
    }

    async saveAsync(order: Order): Promise<void> {
        try {
            const orderJson : string = JSON.stringify(order);
            const values : string[] = [order.uuid, orderJson];
            const script : string = "INSERT INTO Orders (Uuid, Data) VALUES ($1, $2);";
            await this.pool.query(script, values);
        } catch (error) {
            console.error("Error saving order to Postgres:", error);
            throw error;
        } 
    }   
    async getByUuidAsync(uuid: string): Promise<Order | null> {
        try {
            const script : string = "SELECT Data FROM Orders WHERE Uuid = $1;";
            const result = await this.pool.query(script, [uuid]);
            if (result.rows.length === 0) {
                return null;
            }

            const orderData = result.rows[0].data;

            orderData.created_at = new Date(orderData.created_at);
            return orderData as Order;
            
        } catch (error) {
            console.error("Error fetching order by UUID from Postgres:", error);
            throw error;
        } 
    }
}